File: src/lib/tauri/speed.ts
/**
 * Type-safe wrapper around the `change_speed` Tauri command.
 * Dynamically imported to avoid SSR issues (Next.js static export).
 */

async function invoke<T>(cmd: string, args?: Record<string, unknown>): Promise<T> {
  const { invoke: tauriInvoke } = await import("@tauri-apps/api/core");
  return tauriInvoke<T>(cmd, args);
}

export interface ChangeSpeedArgs {
  inputPath: string;
  outputPath: string;
  rate: number;
  hasAudio?: boolean;
  videoCodec?: string;
  crf?: number;
  durationSecs?: number;
  jobId?: string;
}

/** Re-encode a video so it plays at `rate` times the original speed. */
export function changeSpeed(args: ChangeSpeedArgs): Promise<void> {
  return invoke<void>("change_speed", {
    input_path: args.inputPath,
    output_path: args.outputPath,
    rate: args.rate,
    has_audio: args.hasAudio ?? null,
    video_codec: args.videoCodec ?? null,
    crf: args.crf ?? null,
    duration_secs: args.durationSecs ?? null,
    job_id: args.jobId ?? null,
  });
}

File: src-tauri/src/commands/speed.rs
use tauri::AppHandle;

use crate::models::error::AppError;
use crate::services::speed::SpeedService;

/// Re-encode a video so it plays faster or slower.
///
/// `has_audio` defaults to `true`. When `false`, only the video stream is mapped.
#[tauri::command(rename_all = "snake_case")]
pub async fn change_speed(
    app: AppHandle,
    input_path: String,
    output_path: String,
    rate: f64,
    has_audio: Option<bool>,
    video_codec: Option<String>,
    crf: Option<u8>,
    duration_secs: Option<f64>,
    job_id: Option<String>,
) -> Result<(), AppError> {
    if input_path.is_empty() || output_path.is_empty() {
        return Err(AppError::InvalidArgument(
            "input_path and output_path must not be empty".into(),
        ));
    }
    SpeedService::change(
        &app,
        &input_path,
        &output_path,
        rate,
        has_audio.unwrap_or(true),
        video_codec.as_deref(),
        crf,
        duration_secs,
        job_id.as_deref(),
    )
    .await
}
